server/src/main/java/org/apache/uniffle/server/storage/MultiStorageManager.java

```diff
@@ -80,24 +80,43 @@ public boolean write(Storage storage, ShuffleWriteHandler handler, ShuffleDataFl
         if (request == null) {
           return false;
         }
-        storage = warmStorageManager.selectStorage(event);
+        storageManager = storageManager == warmStorageManager ? coldStorageManager : warmStorageManager;
+        event.setStorageManager(storageManager);
+        storage = storageManager.selectStorage(event);
         handler = storage.getOrCreateWriteHandler(request);
       } catch (IOException ioe) {
         LOG.warn("Create fallback write handler failed ", ioe);
         return false;
       }
-      return warmStorageManager.write(storage, handler, event);
+      return storageManager.write(storage, handler, event);
     } else {
       return storageManager.write(storage, handler, event);
     }
   }
 
   private StorageManager selectStorageManager(ShuffleDataFlushEvent event) {
+    if (event.getStorageManager() != null) {
```

Comment:
Could we give a general abstraction for fallback mechanisms. I mean that we can have an fallback interface. And implement two strategies. One is that HDFS can fallback to LOCALFILE, the other is that LOCALFILE can fallback to HDFS. We can choose one with the configuration. And we can expand other strategies if we need.

Reply:
Good idea!

Reply:
good idea